ref-exchange/src/simple_pool.rs

```diff
@@ -63,8 +63,33 @@ impl SimplePool {
         }
     }
 
-    /// Returns
-    pub fn share_balances(&self, account_id: &AccountId) -> Balance {
+    /// Transfers shares from predecessor to receiver.
+    pub fn share_transfer(&mut self, receiver_id: &AccountId, amount: u128) {
+        let account_id = env::predecessor_account_id();
+        self.share_withdraw(&account_id, amount);
+        self.share_deposit(receiver_id, amount);
+    }
+
+    /// Withdraws shares from the given account.
+    pub fn share_withdraw(&mut self, sender_id: &AccountId, amount: u128) {
+        let balance = self.shares.get(&sender_id).expect("ERR_NO_SHARES");
+        if let Some(new_balance) = balance.checked_sub(amount) {
+            self.shares.insert(&sender_id, &new_balance);
+        } else {
+            env::panic(b"ERR_NOT_ENOUGH_SHARES");
+        }
+    }
+
+    /// Deposit shares to the receiver account.
+    pub fn share_deposit(&mut self, receiver_id: &AccountId, amount: u128) {
+        // TODO: add charging for storage.
+        // TODO: handle returns from callback.
+        let balance_out = self.shares.get(&receiver_id).unwrap_or(0);
+        self.shares.insert(&receiver_id, &(balance_out + amount));
+    }
```

Comment:
we don't need withdraw and deposit. We only need `transfer`.

Reply:
It was not clear to me why in FT-141 transfer_call already deposits funds into receiver before the callback resolves.
I would expect funds to be not on the account yet before promise resolves.
But it seems like the standard says we transfer full amount first and return back on callback if needed.

Reply:
The motivation is to have funds, rather than having a promise of having funds.
NOTE: a malicious contract can always do unwanted things with funds - same as  you create an allowance (in case of NEP-21 or ERC-20) - you have no control how the funds will be used unless you audit the contract and all possible updates.